src/System.Composition.Convention/src/Microsoft/Internal/Lock.Reader.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Internal
+{
+    internal struct ReadLock : IDisposable
+    {
+        private readonly Lock _lock;
+        private int _isDisposed;
+
+        public ReadLock(Lock @lock)
+        {
+            _isDisposed = 0;
+            _lock = @lock;
+            _lock.EnterReadLock();
+        }
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
```

Comment:
+1, had a similar thought a couple of hours after that last post :-)

@dsplaisted not sure how the process works; if this is merged and an issue raised for this item I'd love to help out with it. Ping me if/when!